### PR TITLE
Fix token count dtype for fused MoE auxiliary loss which requires int64

### PR DESCRIPTION
**Pull request by zhongbozhu**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Review by yaoyu-33 — APPROVED** [review]

File: megatron/core/transformer/moe/moe_utils.py [hunk 1]
```diff
@@ -280,8 +280,8 @@ def get_tokens_per_expert_and_token_count(
         local_tokens_per_expert, reduce_group
     )
     if with_padding_mask:
-        local_num_tokens = local_tokens_per_expert.sum() / topk
-        total_num_tokens = global_tokens_per_expert.sum() / topk
+        local_num_tokens = local_tokens_per_expert.sum() // topk
```
Comment (yashaswikarnati) [inline megatron/core/transformer/moe/moe_utils.py:283]: just curious, are we assuming they are divisble here ?

Comment (zhongbozhu) [reply megatron/core/transformer/moe/moe_utils.py:283]: That's a good point, tokens per expert received is not necessarily gonna be divisible by topK because that depends on routing. 

Here, TE API only accepts int64, which was the reason why I made this fix. 

```
fused_moe_aux_loss(
    probs,
    tokens_per_expert,
    total_num_tokens,     # Python int OR scalar CUDA torch.int64 tensor
    num_experts,
    topk,
    coeff,               # float; can absorb the normalization adjustment
    deterministic=False,
)
```

If we don't have padding mask, the number of tokens can be directly read from shape, so it's definitely integer. If we do have this padding mask, we need to discuss four cases [THD, BSHD] x [aux_loss, seq_aux_loss]. 

**Assuming TopK=1 BatchSize=2 for BSHD layout:**

For aux_loss and global_aux_loss：

```
Assume seqA has 2 tokens, each token has router probs = [0.9, 0.1], so it selects Expert 0
Assume seqB has 3 tokens, each token has router probs = [0.1, 0.9], so it selects Expert 1

routing_map = [
    [True,  False],  # A0
    [True,  False],  # A1
    [False, True ],  # B0
    [False, True ],  # B1
    [False, True ],  # B2
]

local_tokens_per_expert = routing_map.sum(dim=0)  # tensor([2, 3], dtype=torch.int64), shape [2,]
local_tokens_per_expert.sum() = 5
```

So then 
```
Routing fractions:
f₀ = 2/5 = 0.4
f₁ = 3/5 = 0.6

Average probabilities:
P₀ = (2×0.9 + 3×0.1)/5 = 0.42
P₁ = (2×0.1 + 3×0.9)/5 = 0.58

aux_loss = 2 × (0.4×0.42 + 0.6×0.58)
         = 1.032
```
Here, the 5 is calculated by `total_num_tokens = global_tokens_per_expert.sum() // topk = 5 / 1`, so it's gonna be correct, but you still need the `//` to avoid producing float.  

For seq_aux_loss, we can no longer view the two sequences as a whole, but we want load balancing within each sequences in the batch dimension. 

The correct way:
```
Sequence A:
routing fractions = [1, 0]
average probabilities = [0.9, 0.1]
loss_A = 2 × (1×0.9 + 0×0.1) = 1.8

Sequence B:
routing fractions = [0, 1]
average probabilities = [0.1, 0.9]
loss_B = 2 × (0×0.1 + 1×0.9) = 1.8

seq_aux_loss = ½ × loss_A + ½ × loss_B
             = 1.8
```

Here is what we are doing now, `total_num_tokens = global_tokens_per_expert.sum() / topk  = 5 / (topK * BatchSize) = 5 / (1 * 2) = 2.5` because the `_apply_seq_aux_loss` passed `topk=self.topk * bsz` to `get_tokens_per_expert_and_token_count`, so changing it to integer divide is still wrong. 

Here is the wrong result:
```
Sequence A:
routing fractions     = [2/2.5, 0]           = [0.8, 0]
average probabilities = [2×0.9, 2×0.1]/2.5  = [0.72, 0.08]

loss_A = 2 × (0.8×0.72 + 0×0.08)
       = 1.152                         # should be 1.8

Sequence B:
routing fractions     = [0, 3/2.5]           = [0, 1.2]
average probabilities = [3×0.1, 3×0.9]/2.5  = [0.12, 1.08]

loss_B = 2 × (0×0.12 + 1.2×1.08)
       = 2.592                         # should be 1.8

seq_aux_loss = ½ × 1.152 + ½ × 2.592
             = 1.872                  # should be 1.8
```

**Assuming TopK=1 BatchSize=1 for THD layout:**

We are definitely going to have padding mask, so for both `aux_loss` and `seq_aux_loss`, we trigger `total_num_tokens = global_tokens_per_expert.sum() / topk = 5.0 => a float`.

Comment (yashaswikarnati) [reply megatron/core/transformer/moe/moe_utils.py:283]: Thanks for looking into this in detail. i think this lgtm for the scope of this PR. But we need to handle seq aux loss for THD case properly (may be it is  https://github.com/NVIDIA/Megatron-LM/pull/6115)  and assert token count is an integer.

**Review by yashaswikarnati — APPROVED** [review]

**Comment by yashaswikarnati** [comment]
we will also need #6115 to properly fix seq aux loss for THD

**Review by yaox12 — APPROVED** [review]

**Review by deepakn94 — APPROVED** [review]
